rclcpp_action/include/rclcpp_action/server_goal_handle.hpp

```diff
@@ -15,41 +15,250 @@
+   * This is a terminal state, no more methods should be called on a goal handle after this is
+   * called.
+   * An exception is raised if the goal is in any state besides executing.
+   *
+   * \param[in] result_msg the final result to send to clients.
+   */
+  void
+  set_aborted(typename ACTION::Result::SharedPtr result_msg)
+  {
+    _set_aborted();
+    result_msg->status = action_msgs::msg::GoalStatus::STATUS_ABORTED;
+    on_terminal_state_(uuid_, result_msg);
+  }
 
-  // TODO(sloretz) examples has this attribute as 'goal'
-  /// The original request message describing the goal.
-  const typename ACTION::Goal goal_;
+  /// Indicate that a goal has been reached.
+  /**
+   * Only call this if the goal is executing and has reached the desired final state.
+   * This is a terminal state, no more methods should be called on a goal handle after this is
+   * called.
+   * An exception is raised if the goal is in any state besides executing.
+   *
+   * \param[in] result_msg the final result to send to clients.
+   */
+  void
+  set_succeeded(typename ACTION::Result::SharedPtr result_msg)
+  {
+    _set_succeeded();
+    result_msg->status = action_msgs::msg::GoalStatus::STATUS_SUCCEEDED;
+    on_terminal_state_(uuid_, result_msg);
+  }
+
+  /// Indicate that a goal has been canceled.
+  /**
+   * Only call this if the goal is executing or pending, but has been canceled.
+   * This is a terminal state, no more methods should be called on a goal handle after this is
+   * called.
+   * An exception is raised if the goal is in any state besides executing or pending.
+   *
+   * \param[in] result_msg the final result to send to clients.
+   */
+  void
+  set_canceled(typename ACTION::Result::SharedPtr result_msg)
+  {
+    _set_canceled();
+    result_msg->status = action_msgs::msg::GoalStatus::STATUS_CANCELED;
+    on_terminal_state_(uuid_, result_msg);
+  }
+
+  /// Indicate that the server is starting to execute a goal.
+  /**
+   * Only call this if the goal is pending.
+   * An exception is raised if the goal is in any state besides pending.
+   */
+  void
+  set_executing()
+  {
+    _set_executing();
+    on_executing_(uuid_);
+  }
+
+  /// Get the original request message describing the goal.
+  const std::shared_ptr<const typename ACTION::Goal>
+  get_goal() const
+  {
+    return goal_;
+  }
+
+  /// Get the unique identifier of the goal
+  const std::array<uint8_t, 16>
+  get_uuid() const
+  {
+    return uuid_;
+  }
 
 protected:
-  explicit ServerGoalHandle(const typename ACTION::Goal goal)
-  : goal_(goal) {}
+  /// \internal
+  ServerGoalHandle(
+    std::shared_ptr<rcl_action_server_t> rcl_server,
+    std::shared_ptr<rcl_action_goal_handle_t> rcl_handle,
+    std::array<uint8_t, 16> uuid,
+    std::shared_ptr<const typename ACTION::Goal> goal,
+    std::function<void(const std::array<uint8_t, 16>&, std::shared_ptr<void>)> on_terminal_state,
+    std::function<void(const std::array<uint8_t, 16>&)> on_executing
```

Comment:
Can we have an alias for `std::array<uint8_t, 16>`? like `rclcpp_action::GoalID` or whatever?

Reply:
Sorry, just saw https://github.com/ros2/rclcpp/pull/593#discussion_r238888237

Reply:
Definitely. It's in the client PR. I'm hoping to update this one after the client is merged.

Reply:
Done in 
b9e7fc8